hashes/src/sha256.rs

```diff
@@ -114,33 +120,41 @@ impl Hash {
     /// Computes hash from `bytes` in `const` context.
     ///
     /// Warning: this function is inefficient. It should be only used in `const` context.
-    pub const fn const_hash(bytes: &[u8]) -> Self { Hash(Midstate::const_hash(bytes, true).0) }
+    pub const fn const_hash(bytes: &[u8]) -> Self { Hash(Midstate::const_hash(bytes, true).bytes) }
 }
 
 /// Output of the SHA256 hash function.
 #[derive(Copy, Clone, PartialEq, Eq, Default, PartialOrd, Ord, Hash)]
-pub struct Midstate(pub [u8; 32]);
+pub struct Midstate {
+    /// Raw bytes of the midstate i.e., the already-hashed contents of the hash engine.
+    bytes: [u8; 32],
+    /// Number of bytes hashed to achieve this midstate.
+    // INVARIANT must always be a multiple of 64.
+    length: u64,
+}
 
 impl Midstate {
-    /// Length of the midstate, in bytes.
-    const LEN: usize = 32;
-
     /// Construct a new [`Midstate`] from the inner value.
-    pub const fn from_byte_array(inner: [u8; 32]) -> Self { Midstate(inner) }
+    pub const fn new(state: [u8; 32], length: u64) -> Self {
+        Midstate {
```

Comment:
Should panic if length is not multiple of 64.

Reply:
Ah the whole Pr does not enforce the invariant, that was intentional, I should have explained that explicitly.

Reply:
Sorry man, I feel like I abused your time by pushing this up without a better explanation of the state, I'll try to be better.